### PR TITLE
Updated elevation angle field names to support FitACF 3.0

### LINKED ISSUE #357: Elv_low Elv_high in Fitacf 3.0
So I left the `elv_low` and `elv_high` the same as Pasha implemented them as (not the same as FITACF 2.5, @ecbland if you can explain better please do so!) in PR #356; however, because they have 2 meanings based on the algorithm picked and may not be aware to the user that the meanings have changed between versions. 

I am wondering if we want to convert FITACF 3.0 `elv_low` and `elv_high` back to the meaning of what FITACF 2.5 set it as and ask DSWG about creating 2 new FITACF fields for FITACFs 3.0 current `elv_low` and `elv_high`?

Please do not ask me specifics as I don't really know much except what is in the code.  However, suggestive ideas and solutions are most encouraged and what actions to take next on this given problem.  

-----

EDIT by @ecbland :

FITACF 2.5 (documented [here](https://radar-software-toolkit-rst.readthedocs.io/en/latest/references/general/fitacf/))
- `elv` is the elevation angle calculated from the fitted phase
- `elv_high` is the upper estimate of the elevation angle
- `elv_low` is the lower estimate of the elevation angle

FITACF 3.0 (documented [here](https://superdarn.github.io/dawg/files/sup_material/FITACF3_release_notes.pdf))
- `elv` is the elevation angle calculated from the lag zero phase of the ACF
- `elv_high` is the elevation angle calculated from the fitted phase (cf. FITACF 2.5 `elv`)
- `elv_low` is the least square error

### PR DESCRIPTION
**Pull request by ecbland**
This PR introduces two new fields to the `fitacf` file format: `elv_fitted` and `elv_error` (see [February 2021 DAWG report](https://superdarn.github.io/dawg/reports/202102/#fitacf3-elevation-fields), closes #357). This change was approved by the PIEC in November 2021. 

Additionally, I decided to experiment by completely removing the XCF arrays that FitACF 3.0 does not populate (`x_v`, `x_p_l`, etc.) Previously these arrays were just filled with zeros, but now they are completely removed. **Please let me know if this is a good idea!**

### Testing

Run `make_fit -fitacf-version 3.0` on any `rawacf` file that includes interferometer data (`prm.xcf=1`). Then use `dmapdump -d` to view the new field names and their values. 

You can preview a sample file created on this branch here: 
[modified_format.lyr.fitacf3.zip](https://github.com/SuperDARN/rst/files/8002959/modified_format.lyr.fitacf3.zip)

**I would appreciate your help with checking that I have included all the fields that should be there, and none that shouldn't be there!**

**Checklist:** 
- [x] No change to any behaviour when using FitACF 2.5
- [x] Can read the new `fitacf3` files with IDL DLMs
- [x] Can read the new `fitacf3` files with the native IDL code
- [x] Can still read any `fitacf` file produced on `develop` branch
- [x] Works in pyDARNio

### Notes 

1. In our next DAWG report I will ask the PIs to review this change and provide feedback. 
2. We will make this update in parallel with pyDARNio (see https://github.com/SuperDARN/pyDARNio/issues/38 and https://github.com/SuperDARN/pyDARNio/pull/39)

### CODE REVIEW

File: codebase/superdarn/src.idl/lib/main.1.25/fit.pro [hunk 1]
```diff
@@ -255,11 +261,24 @@ function FitRead,unit,prm,fit
...
-      fit.x_sd_phi[slist]= (*(arrvec[arrid[37]].ptr))[*]
+      fit.elv_low[slist]= (*(arrvec[arrid[35]].ptr))[*]
+      fit.elv_high[slist]= (*(arrvec[arrid[36]].ptr))[*]
+      fit.x_sd_l[slist]= (*(arrvec[arrid[37]].ptr))[*]
+      fit.x_sd_s[slist]= (*(arrvec[arrid[38]].ptr))[*]
+      fit.x_sd_phi[slist]= (*(arrvec[arrid[39]].ptr))[*]
+    endif
+    
+    ; XCF fitted parameters for FitACF 3
+    if (fit.revision.major ge 3) then begin
```
Comment (egthomas) [inline codebase/superdarn/src.idl/lib/main.1.25/fit.pro:272]: @ecbland this value was `5` from ~2004-2019 (and `4` from ~2000-2004), so there may be issues with files generated using previous versions of FITACF / RST / ROS: 

6dd7dd348dde3ad12174a313147f1c4daa0bbb4a
https://github.com/SuperDARN/rst-archive/blob/rst.2.05/code/src.lib/superdarn/fitacf.1.22/src/fitacfversion.h
https://github.com/SuperDARN/ros-archive/blob/ros.0.93A/qnx/code/src.bin/superdarn/process/make_fit.1.04/make_fit.c#L141

**Comment by pasha-ponomarenko** [comment]
I tested the new branch with 20180404.0801.00.inv.rawacf.

This file has been processed with make_fit -fitacf-version 3.0 etc using the current default package (rst-release-rst-4.6) and the new branch (rst-ehn-fitacf3-elevation-fields).

;----------------------------------------------------------------------------------------------
FIT structure:

The old FIT structure contains 40 fields with 15 related to XXCF power fitting. 
The new FIT structure contains 26 fields. 
All fields with 'x_' prefix except x_sd_phi (stddev for phase fit) were removed as expected.

;----------------------------------------------------------------------------
PRM structure
	
PRM structures in both cases contain 51 fields of identical content as exected.

;-----------------------------------------------------------------------------
Data completeness/correspondence:

In order to make sure that there is one-to-one correspondence between the new and the old elevation-related fields, I checked the sums and differences between the corresponding fields in the FIT structures:

IDL> print, total(fit_new.elv_fitted), total(fit_old0.elv_high), total(fit_old0.elv_high - fit_new.elv_fitted)
      392822.      392822.      0.00000
IDL> print, total(fit_new.elv_error), total(fit_old0.elv_low), total(fit_old0.elv_low - fit_new.elv_error)
      12101.3      12101.3      0.00000
IDL> print, total(fit_new.elv), total(fit_old0.elv), total(fit_old0.elv - fit_new.elv)
      393457.      393457.      0.00000
      
They show that the re-naming the fields worked as expected.

;---------------------------------------------------------------------------------

**Comment by pasha-ponomarenko** [comment]
I was also able to read "old" FITACF files with "new" IDL DLMs, and vice versa. The "alien" fields remained empty, as one might expect.

File: codebase/superdarn/src.idl/lib/main.1.25/fit.pro (hunk 1 above)
Comment (ecbland) [reply codebase/superdarn/src.idl/lib/main.1.25/fit.pro:272]: @egthomas Thanks for reminding me about this :scream: 

I suppose the simplest solution is to change those lines to `if (fit.revision.major eq 3)`, and then if someone ever writes a FitACF 4.0 they can find a different solution. I can leave some comments in the code as well. Does that sound sensible?

**Comment by mts299** [comment]
Looks good to me, fields are different in fitacf 3.0 files and seems the same from dev and fitacf 2.5 is different.

**Comment by mts299** [comment]
This is not be merged until after the next release because of the major change to the file. 

@ecbland did you update the FITACF documentation in readthedocs on the changes?

**Comment by ecbland** [comment]
> did you update the FITACF documentation in readthedocs on the changes?

Yes, the updated documentation is part of this PR. https://github.com/SuperDARN/rst/pull/472/files#diff-873897527e7ff2ef7be874e432de67656215cfad325d583b15cc52ac65b4df75

**Comment by egthomas** [comment]
@ecbland I'm not sure where this currently stands but the merge conflicts should now be resolved.

**Comment by ecbland** [comment]
Thanks for fixing the conflicts, @egthomas! I'd say this is ready to merge.

**Comment by ecbland** [comment]
This PR is ready to merge! The necessary updates to `pyDARNio` have been merged (https://github.com/SuperDARN/pyDARNio/pull/39), and I have resolved the merge conflicts on this branch (comment lines only, introduced by #524) and tested one last time. 

Could someone please press the big green button if you agree this work is finished?

**Comment by egthomas** [comment]
@ecbland I know this has been open for a while already, but would it be worth using the new fitting algorithm field to identify files created with FITACF 3.0 rather than `fit.revision.major`?  I know there are some obvious downsides (eg files that were produced prior to the `develop` branch / upcoming RST release) but it may be a more stable long-term solution.  Just a thought.

**Comment by pasha-ponomarenko** [comment]
@egthomas, I wonder if this is a matter for a separate pull request?

**Comment by ecbland** [comment]
> Then I guess I can press the green button...

@pasha-ponomarenko Yes please!

**Pull request merged**